tool/tbot/identity/identity.go

```diff
@@ -0,0 +1,388 @@
+	certSigner, err := ssh.NewCertSigner(cert, signer)
+	if err != nil {
+		return nil, trace.BadParameter("unsupported private key: %v", err)
+	}
+
+	// check principals on certificate
+	if len(cert.ValidPrincipals) < 1 {
+		return nil, trace.BadParameter("valid principals: at least one valid principal is required")
+	}
+	for _, validPrincipal := range cert.ValidPrincipals {
+		if validPrincipal == "" {
+			return nil, trace.BadParameter("valid principal can not be empty: %q", cert.ValidPrincipals)
+		}
+	}
+
+	// check permissions on certificate
+	// TODO: do we care to even verify this at all?
+	// if len(cert.Permissions.Extensions) == 0 {
+	// 	return nil, trace.BadParameter("extensions: misssing needed extensions for host roles")
+	// }
+
+	// roleString := cert.Permissions.Extensions[utils.CertExtensionRole]
+	// if roleString == "" {
+	// 	return nil, trace.BadParameter("misssing cert extension %v", utils.CertExtensionRole)
+	// }
+	// roles, err := types.ParseTeleportRoles(roleString)
+	// if err != nil {
+	// 	return nil, trace.Wrap(err)
+	// }
+	// foundRoles := len(roles)
+	// if foundRoles != 1 {
+	// 	return nil, trace.Errorf("expected one role per certificate. found %d: '%s'",
+	// 		foundRoles, roles.String())
+	// }
+
+	// TODO: host certs use CertExtensionAuthority, client certs use CertExtensionTeleportRouteToCluster
+	// (or at least, teleport-route-to-cluster _appears_ to be a sane cluster name, but I have no idea
+	// how it is actually set.)
+	//clusterName := cert.Permissions.Extensions[utils.CertExtensionAuthority]
+	clusterName := cert.Permissions.Extensions[teleport.CertExtensionTeleportRouteToCluster]
+	if clusterName == "" {
+		return nil, trace.BadParameter("missing cert extension %v", utils.CertExtensionAuthority)
+	}
+
+	return &Identity{
+		ClusterName:       clusterName,
+		KeyBytes:          keyBytes,
+		SSHPublicKeyBytes: publicKeyBytes,
+		CertBytes:         certBytes,
+		KeySigner:         certSigner,
+		Cert:              cert,
+	}, nil
+}
+
+func SaveIdentity(id *Identity, d destination.Destination) error {
+	for _, data := range []struct {
+		name     string
+		data     []byte
+		modeHint destination.ModeHint
+	}{
+		{TLSCertKey, id.TLSCertBytes, destination.ModeHintSecret},
+		{SSHCertKey, id.CertBytes, destination.ModeHintSecret},
+		{TLSCACertsKey, bytes.Join(id.TLSCACertsBytes, []byte("$")), destination.ModeHintSecret},
+		{SSHCACertsKey, bytes.Join(id.SSHCACertBytes, []byte("\n")), destination.ModeHintSecret},
+		{PrivateKeyKey, id.KeyBytes, destination.ModeHintSecret},
+		{PublicKeyKey, id.SSHPublicKeyBytes, destination.ModeHintUnspecified},
+	} {
+		log.Debugf("Writing %s", data.name)
+		if err := d.Write(data.name, data.data, data.modeHint); err != nil {
+			return trace.Wrap(err, "could not write to %v", data.name)
+		}
+	}
+	return nil
+}
+
+func LoadIdentity(d destination.Destination) (*Identity, error) {
+	// TODO: encode the whole thing using the identityfile package?
```

Comment:
It does look like there's a lot of functionality overlap tbh. Any way to reuse some of the identityfile-provided functions?